segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/DirectMemoryBuffer.java

```diff
@@ -0,0 +1,493 @@
+                // unreferenced data in the buffer.
+                rollbackWrite(this.layout.getBlockId(predecessorAddress), blockId);
+            }
+
+            throw ex;
+        }
+
+        // Update the root metadata.
+        blockMetadata = metadataBuf.getLong(0);
+        blockMetadata = this.layout.setNextFreeBlockId(blockMetadata, blockId);
+        metadataBuf.setLong(0, blockMetadata);
+
+        return new WriteResult(dataOffset, predecessorAddress, firstBlockId);
+    }
+
+    /**
+     * Attempts to append some data to an entry.
+     *
+     * @param blockId        The id of the Buffer-Block to append that resides in this buffer.
+     * @param expectedLength The expected length of the Buffer-Block to append to. If this value does not match what is
+     *                       stored in that Buffer-Block metdata, the append will not execute.
+     * @param data           A {@link BufferView} representing the data to append. The {@link BufferView#copyTo(ByteBuffer)}
+     *                       method will be used to transfer the data to the appropriate Buffer Block.
+     * @return The number of bytes appended.
+     * @throws IncorrectCacheEntryLengthException If `expectedLastBlockLength` differs from the last block's length.
+     * @throws IllegalArgumentException If blockId does not point to an allocated Block.
+     */
+    synchronized int tryAppend(int blockId, int expectedLength, BufferView data) {
+        validateBlockId(blockId, false);
+
+        ByteBuf metadataBuf = getMetadataBlock();
+        int bufIndex = blockId * this.layout.blockMetadataSize();
+        long blockMetadata = metadataBuf.getLong(bufIndex);
+        Preconditions.checkArgument(blockId != CacheLayout.NO_BLOCK_ID && this.layout.isUsedBlock(blockMetadata),
+                "Given blockId is not allocated.");
+
+        // Validate that the given length matches the actual one.
+        int blockLength = this.layout.getLength(blockMetadata);
+        if (blockLength != expectedLength) {
+            throw new IncorrectCacheEntryLengthException(String.format(
+                    "Incorrect last block length. Expected %s, given %s.", blockLength, expectedLength));
+        }
+
+        // Adjust the length, if we were given more than we can fit.
+        int maxLength = this.layout.blockSize() - blockLength;
+        if (maxLength < data.getLength()) {
+            data = data.slice(0, maxLength);
+        }
+
+        // Copy the data.
+        data.copyTo(getWriteableBlock(blockId, blockLength));
+        blockLength += data.getLength();
+
+        // Update metadata.
+        blockMetadata = this.layout.setLength(blockMetadata, blockLength);
+        metadataBuf.setLong(bufIndex, blockMetadata);
+        return data.getLength();
+    }
+
+    /**
+     * Reads from the buffer starting at the given Buffer-Block Id and moving backwards (following
```

Comment:
This is a very odd way to describe the API it is "moving backwars" but returns the items in reverse order. Would it be still accurate to say, it reads all of the data and returns it in order?

Reply:
Rephrased.